### ISSUE TITLE
ShyLU Basker build failure with new CMake changes

### ISSUE THREAD

**Issue by ndellingwood** [trilinos/Trilinos#2075]
While testing the recent CMake changes discussed in issue #1400 Trilinos fails to build with ShyLU Basker enabled with the following error:
```
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/amesos2/src/Amesos2_Basker_FunctionMap.hpp:64:41: fatal error: shylubasker_trilinos_decl.hpp: No such file or directory
compilation terminated.
```
The file above is not being found from the directory `Trilinos/packages/shylu/shylu_node/basker/src`

@trilinos/kokkos @trilinos/shylu @bartlettroscoe @crtrott @ndellingwood 

## Current Behavior

With the new CMake changes Trilinos fails to build with ShyLU Basker enabled. 

## Motivation and Context

Trilinos builds with ShyLU Basker enabled on the develop branch without the CMake changes. Xyce uses this package and would be affected by this issue. 

## Steps to Reproduce

1. Clone Kokkos and checkout the develop branch
2. Clone Trilinos and checkout the kokkos-promotion branch
3. Create symbolic link to kokkos in the Trilinos directory; add corresponding line to configure line/file: `-DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos`
4. Create a build directory and configure Trilinos - my configure script is attached (change the TPL directories if using system installs or SEMS modules)
5. Build Trilinos - make -j 
6. Observe the error towards the end of compilation ~95%

Configure script:
```
#!/bin/bash

export SRCDIR=/Users/ndellin/Research/trilinos-dev/Trilinos

TPL_PATH=/Users/ndellin/Research/TPL
BLAS_PATH=${TPL_PATH}/BLAS-3.6.0
LAPACK_PATH=${TPL_PATH}/lapack-3.6.1
SCOTCH_PATH=${TPL_PATH}/scotch_6.0.3

export INSTALL_LOCATION=${PWD}/install

rm -rf CMakeCache.txt CMakeFiles

cmake \
\
-D CMAKE_INSTALL_PREFIX:PATH="${INSTALL_LOCATION}" \
-D CMAKE_CXX_FLAGS:STRING="-DSHYLU_NODEBASKER -g" \
-D CMAKE_MAKE_PROGRAM="make" \
-D CMAKE_BUILD_TYPE:STRING=RELEASE \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
-D BUILD_SHARED_LIBS:BOOL=OFF \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=OFF \
\
-D TPL_ENABLE_MPI:BOOL=ON \
  -D CMAKE_CXX_COMPILER:FILEPATH="`which mpicxx`" \
  -D CMAKE_C_COMPILER:FILEPATH="`which mpicc`" \
  -D CMAKE_Fortran_COMPILER:FILEPATH="`which mpifort`" \
\
-D TPL_ENABLE_ParMETIS:BOOL=OFF \
-D TPL_ENABLE_Scotch:BOOL=ON \
  -D Scotch_INCLUDE_DIRS:PATH="${SCOTCH_PATH}/include" \
  -D Scotch_LIBRARY_DIRS:PATH="${SCOTCH_PATH}/lib" \
  -D Scotch_LIBRARY_NAMES:STRING="scotch;scotcherr" \
-D TPL_ENABLE_BLAS:STRING=ON \
  -D BLAS_LIBRARY_DIRS:FILEPATH=${BLAS_PATH} \
  -D TPL_BLAS_LIBRARIES="-lblas" \
-D TPL_ENABLE_LAPACK:STRING=ON \
  -D LAPACK_LIBRARY_DIRS:FILEPATH=${LAPACK_PATH} \
  -D TPL_LAPACK_LIBRARIES="-llapack" \
\
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON \
-D Trilinos_ENABLE_DEBUG:BOOL=OFF \
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
\
-D Trilinos_ENABLE_Amesos:BOOL=ON \
  -D Amesos_ENABLE_KLU:BOOL=ON \
\
-D Trilinos_ENABLE_CHECKED_STL:BOOL=OFF \
-D Trilinos_ENABLE_OpenMP:BOOL=ON \
-D Trilinos_ENABLE_Kokkos:BOOL=ON \
-D Kokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
  -D Kokkos_ENABLE_OpenMP:BOOL=ON \
  -D Kokkos_ENABLE_EXAMPLES:BOOL=OFF \
  -D Kokkos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-D Trilinos_ENABLE_Epetra:BOOL=ON \
  -D Epetra_ENABLE_EXAMPLES:BOOL=OFF \
  -D Epetra_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EpetraExt:BOOL=ON \
  -D EpetraExt_ENABLE_EXAMPLES:BOOL=OFF \
  -D EpetraExt_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_Tpetra:BOOL=ON \
  -D Tpetra_ENABLE_EXAMPLES:BOOL=OFF \
  -D Tpetra_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_TeuchosRemainder:BOOL=ON \
-D Trilinos_ENABLE_ShyLU_NodeBasker:BOOL=ON \
  -D ShyLU_NodeBasker_ENABLE_TESTS:BOOL=ON \
  -D ShyLU_NodeBasker_ENABLE_EXAMPLES:BOOL=ON \
-D Trilinos_ENABLE_Amesos2:BOOL=ON \
  -D Amesos2_ENABLE_TESTS:BOOL=ON \
  -D Amesos2_ENABLE_EXAMPLES:BOOL=ON \
  -D Amesos2_ENABLE_TIMERS:BOOL=ON \
  -D Amesos2_ENABLE_KLU2:BOOL=ON \
  -D Amesos2_ENABLE_Basker:BOOL=ON \
${SRCDIR}
```

## Your Environment

Tested on laptop with cmake/3.6.2, gcc/5.4.0
TPLs required: BLAS, LAPACK, Scotch 6.0 (available as SEMS module sems-scotch/6.0.3/nopthread_64bit_parallel)
- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 

## Additional Information

Compile line preceding the error:
`cd /Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/amesos2/src && /opt/local/bin/mpicxx -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest -I/Users/ndellin/Research/trilinos-dev/Trilinos -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/amesos2/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/amesos2/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/amesos2/src/KLU2/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/amesos2/src/KLU2/Source -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/AMD/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/COLAMD/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/BTF/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/UFconfig -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/epetraext/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/transform -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/inout -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/coloring -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/model_evaluator -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/block -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetraext/src/restrict -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/epetra/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/epetra/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/teuchos/kokkoscomm/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/kokkoscomm/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/teuchos/kokkoscompat/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/kokkoscompat/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/parameterlist/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/parser/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/teuchos/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/kokkos/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/kokkos/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/comm/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/remainder/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/teuchos/remainder/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/teuchos/numerics/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/CAMD/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/CCOLAMD/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/common/auxiliarySoftware/SuiteSparse/src/KLU/Include -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/core/ext -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/core/ext -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/core/inout -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/core/inout -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/core/src/kokkos_refactor -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/core/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/tsqr/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/tsqr/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/classic/LinAlg -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/classic/NodeAPI -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/classic/NodeAPI -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/tpetra/classic/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/tpetra/classic/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/packages/kokkos-kernels/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/impl -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/impl/tpls -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/blas -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/blas/impl -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/sparse -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/sparse/impl -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/graph -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/graph/impl -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/batched -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/batched/impl -I/Users/ndellin/Research/trilinos-dev/Trilinos/packages/kokkos-kernels/src/common -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/kokkos/algorithms/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/kokkos/algorithms/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/Build/ShyLUBasker/CMakeTest/kokkos/containers/src -I/Users/ndellin/Research/trilinos-dev/Trilinos/kokkos/containers/src -pedantic -Wall -Wno-long-long -Wwrite-strings -DSHYLU_NODEBASKER -DBASKER_TIMER -g -std=c++11 -fopenmp -O3 -DNDEBUG -o CMakeFiles/amesos2.dir/Amesos2_Factory.cpp.o -c /Users/ndellin/Research/trilinos-dev/Trilinos/packages/amesos2/src/Amesos2_Factory.cpp`

**Comment by bartlettroscoe** [comment]
@ndellingwood,

Why does the compiler define `-DSHYLU_NODEBASKER ` need to be manually set in the `CMAKE_CXX_FLAGS`?  Should not not get set automatically using some CMake logic like described in:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-se-package-dependency

?

**Comment by ndellingwood** [comment]
@bartlettroscoe I think that was done to distinguish between an existing serial implementation of Basker in Amesos2 (originally intended to be temporary) and the threaded impl in ShyLU.

**Comment by ndellingwood** [comment]
`SHYLU_NODEBASKER` is recognized during configuration/compilation (or else the file not found above would be if-guarded out), but the path to file is not found.

**Comment by ndellingwood** [comment]
@bartlettroscoe should the current configuration setup work with the new CMake system? If not we can decide on best way to reconcile the two Basker versions (different names,...), it's been on the radar but can be bumped higher in priority.

**Comment by bartlettroscoe** [comment]
@ndellingwood,

I found the problem.  The problem is that the `ShyLU_NodeBasker` subpackage is not defining any libraries in this case and therefore does not have any include directories to export.  The problem is that the new Kokkos system does not define and turn on a CMake variable called `KOKKOIS_HAVE_OPENMP`.  

You can see this in the file:

   * Trilinos/packages/shylu/shylu_node/basker/CMakeLists.txt

which is:

```
TRIBITS_SUBPACKAGE(Basker)

IF(KOKKOS_HAVE_OPENMP)
      
  ADD_SUBDIRECTORY(src)

  TRIBITS_ADD_TEST_DIRECTORIES(test)

ENDIF()

#TRIBITS_ADD_EXAMPLE_DIRECTORIES(example)

TRIBITS_SUBPACKAGE_POSTPROCESS()
```

Instead of just skipping the entire subpackage, would it not be better to just assert and require that Kokkos is configured with OpenMP support in order to configure the ShyLU_NodeBasker subpackage?   Or could we put in special logic to disable ShyLU_NodeBasker if OpenMP support is turned off (if possible to know that before Kokkos package is actually processed)?

Just silently skipping the entire subpackage targets and tests seems like a bad way to do this.

At the very least, this CMakeLists.txt file should use the macro `ASSERT_DEFINED()` like:

```
TRIBITS_SUBPACKAGE(Basker)

ASSERT_DEFINED(KOKKOS_HAVE_OPENMP)
IF (KOKKOS_HAVE_OPENMP)
      
  ADD_SUBDIRECTORY(src)

  TRIBITS_ADD_TEST_DIRECTORIES(test)

ENDIF()

#TRIBITS_ADD_EXAMPLE_DIRECTORIES(example)

TRIBITS_SUBPACKAGE_POSTPROCESS()
```

That way, if the variable `KOKKOS_HAVE_OPENMP` is not even defined, it will error out.  I added that to my local branch and it prints a very nice error message:

```
Processing enabled package: ShyLU_Node (Basker, Tests, Examples)
CMake Error at cmake/tribits/core/utils/AssertDefined.cmake:79 (MESSAGE):
  Error, the variable KOKKOS_HAVE_OPENMP is not defined!
Call Stack (most recent call first):
  packages/shylu/shylu_node/basker/CMakeLists.txt:3 (ASSERT_DEFINED)
```

I can commit that to the branch if you like and then we can decide how to fix this.

With your permission, I can add the `ASSERT_DEFINED()` to the branch and then we can decide how to address this.

**Comment by bartlettroscoe** [comment]
On a related note, in TriBITSPub/TriBITS#200, we have been trying to add checks to TriBITS to catch as many user mistakes like this as we can.  I am wondering if we can add an automated check to TriBITS to help catch problems like this.

Should TriBITS try to catch mistakes like this and not allow a package or subpackage to not have any libraries or tests or examples?  We could add a check for TriBITS that would disallow this.  A package with no libraries, tests, or examples seems like it serves no purpose for compiled code.  But I could see a use case where a TriBITS package or subpackage just provided some scripts or something and then using the TriBITS package architecture could be used to make sure that such packages ended up in tarballs for downstream packages that need them.  So we could allow such packages to explicitly ask skip that check.  But the default would be to not allow that.  In fact, by default, we may not even allow packages to not have any libraries, unless they specifically state that they don't have any libraries.  That would have caught this mistake as well.

I am just trying to think of a way that TriBITS could help catch mistakes like this before they become a problem down the line like this.

**Comment by ndellingwood** [comment]
@bartlettroscoe thanks for the suggestions, I just discovered the HAVE vs ENABLE macro issue while rewriting the Amesos2 ShyLUBasker interface to properly make use Tribits. 

> Instead of just skipping the entire subpackage, would it not be better to just assert and require that Kokkos is configured with OpenMP support in order to configure the ShyLU_NodeBasker subpackage
Yes, if someone is trying to use ShyLUBasker then they should be using Kokkos and have OpenMP enabled. 

> I can commit that to the branch if you like and then we can decide how to fix this.
Yes please, if that seems the fastest way to get past this for now. 

> The problem is that the new Kokkos system does not define and turn on a CMake variable called `KOKKOS_HAVE_OPENMP`
Just to clarify, from here on should all `KOKKOS_HAVE_*` be replaced by `KOKKOS_ENABLE_*`? This may be related to the Intrepid2 issues documented in #1928, it seems like some configure info (lambda-related) may not be percolating its way to the code correctly, but I can't verify that with certainty yet.

**Comment by bartlettroscoe** [comment]
@ndellingwood,

> Yes, if someone is trying to use ShyLUBasker then they should be using Kokkos and have OpenMP enabled.

I think that means that we can put logic into the file:

* Trilinos/cmake/RepositoryDependenciesSetup.cmake

that will disable ShyLU_NodeBasker if KokkosCore(?) and OpenMP are not both enabled.  And it could issue an error right there if someone tries to enable ShyLU_NodeBasker when these others are not set.

Does that sounds like a good way to handle this?

> Yes please, if that seems the fastest way to get past this for now.

Okay, I will do that right now.

> Just to clarify, from here on should all `KOKKOS_HAVE_*` be replaced by `KOKKOS_ENABLE_*?`

TriBITS technically does not mandate the names of these variables (that are translated into C/C++ defines in the generated XXX_config.h files) but it goes against the TriBITS convention to do this.  See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-se-package-dependency
* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

**Comment by mhoemmen** [comment]
Kokkos uses `KOKKOS_ENABLE_*` as CMake option names now.  It may retain `KOKKOS_HAVE_*` equivalents for backwards compatibility.

**Comment by ndellingwood** [comment]
I pushed changes to kokkos-promotion branch changing all occurrences of `KOKKOS_HAVE_*` with `KOKKOS_ENABLE_*` which fixed the issue I was seeing, and Ross's `ASSERT_DEFINED` addition now checks for this issue; working on a proper Amesos2 interface for ShyLUBasker now, but will open up a separate issue and PR for that once complete (and after the integration is finished). 
Will wait to close until deciding on if adding cmake logic for the dependencies should be added to consider this complete or opened as a separate issue.

**Comment by bartlettroscoe** [comment]
> Will wait to close until deciding on if adding cmake logic for the dependencies should be added to consider this complete or opened as a separate issue.

The logic is easy to add. But given that ShyLU_NodeBasker is currently listed as an `EX` package, it is already disabled by default no one is going to accidentally enable it.  But rather than just silently dropping out the entire implementation of ShyLU_NodeBasker (which the current implementation is still doing),  which has very bad consequences downstream (like reported in this issue), at the very least, the ShyLU_NodeBasker CMakeLists.txt file should be changed to:

```
TRIBITS_SUBPACKAGE(Basker)

ASSERT_DEFINED(KOKKOS_ENABLE_OPENMP)
IF (NOT KOKKOS_ENABLE_OPENMP)
  MESSAGE(FATAL_ERROR
    "Error, can't enable ${PACKAGE_NAME} if 'KOKKOS_ENABLE_OPENMP='${KOKKOS_ENABLE_OPENMP}'"
    " is not 'ON'!")
ENDIF()
      
ADD_SUBDIRECTORY(src)

TRIBITS_ADD_TEST_DIRECTORIES(test)

#TRIBITS_ADD_EXAMPLE_DIRECTORIES(example)

TRIBITS_SUBPACKAGE_POSTPROCESS()
```

That solves this problem until this subpackage is elevated to `ST` or `PT`.  At that point,the above implementation is not really acceptable.
